README.md
# (MODX)EvolutionCMS.snippets.ddTypograph

Snippet for text typography.
The snippet doesn’t use third-party services, also it sends no requests. In other words, everything is performed on your server.


## Requires

* PHP >= 5.6
* [(MODX)EvolutionCMS.libraries.ddTools](https://code.divandesign.biz/modx/ddtools) >= 0.48.1
* [PHP.libraries.EMT](http://mdash.ru) 3.5 (contains in archive)


## Documentation


### Installation


#### 1. Elements → Snippets: Create a new snippet with the following data

1. Snippet name: `ddTypograph`.
2. Description: `<b>2.5</b> Snippet for text typography. The snippet doesn’t use third-party services, also it sends no requests. In other words, everything is performed on your server.`.
3. Category: `Core`.
4. Parse DocBlock: `no`.
5. Snippet code (php): Insert content of the `ddTypograph_snippet.php` file from the archive.


#### 2. Elements → Manage Files

1. Create a new folder `assets/snippets/ddTypograph/`.
2. Extract the archive to the folder (except `ddTypograph_snippet.php`).


### Parameters description

* `text`
	* Desctription: Text to correct.
	* Valid values: `string`
	* **Required**
	
* `optAlign`
	* Desctription: Optical alignment (hanging punctuation).
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `text_paragraphs`
	* Desctription: Section signs and line breaks insertion.
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `text_autoLinks`
	* Desctription: Marking links (including email ones).
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `etc_unicodeConvert`
	* Desctription: Convert HTML entities into Unicode (`—` instead of `&mdash;`, etc).
	* Valid values:
		* `0`
		* `1`
	* Default value: `1`
	
* `noTags`
	* Desctription: Whether HTML element insertion is allowed or not.  
		There are cases when using tags causes the text to be invalid, for example, using the snippet inside of an HTML attribute.
	* Valid values:
		* `0`
		* `1`
	* Default value: `0`
	
* `excludeTags`
	* Desctription: HTML tags which content will be ignored by snippet.
	* Valid values: `stringCommaSeparated`
	* Default value: `'notg,code'`


### Examples


#### Typography the content

```
[[ddTypograph? &text=`[*content*]`]]
```


#### Typography text with auto paragraphs, links & emails

```
[[ddTypograph?
	&text=`Some text for typography.`
	&text_paragraphs=`1`
	&text_autoLinks=`1`
]]
```


#### Typography text with auto optical margin alignment (hanging quotes & etc)

```
[[ddTypograph?
	&text=`Some text for "typography".`
	&optAlign=`1`
]]
```


#### Turn off typography for a text fragment (the `<notg></notg>` tag)

```html
[[ddTypograph?
	&text=`Some text. <notg>The snippet will not change this text inside the tag.</notg> It's easy.`
]]
```


#### Run the snippet through `\DDTools\Snippet::runSnippet` without DB and eval

```php
\DDTools\Snippet::runSnippet([
	'name' => 'ddTypograph',
	'params' => [
		'text' => "
			There's nothing you can do that can't be done
			Nothing you can sing that can't be sung
			
			https://en.wikipedia.org/wiki/The_Beatles
		",
		'text_paragraphs' => true,
		'text_autoLinks' => true
	]
]);
```

Returns:

```
<p>There’s nothing you can do&nbsp;that can’t be&nbsp;done<br>
Nothing you can sing that can’t be&nbsp;sung</p>
<p><a href="https://en.wikipedia.org/wiki/The_Beatles">en.wikipedia.org/wiki/The_Beatles</a></p>
```


## Links

* [Home page](https://code.divandesign.biz/modx/ddtypograph)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-snippets-ddtypograph)


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.snippets.ddTypograph changelog


## Version 2.5 (2021-03-29)
* \* Attention! PHP >= 5.6 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.48.1 is required.
* \+ You can just call `\DDTools\Snippet::runSnippet` to run the snippet without DB and eval (see README → Examples).
* \+ `\ddDomainComparator\Snippet`: The new class. All snippet code was moved here.
* \+ README → Links.
* \* Composer.json:
	* \+ `homepage`
	* \+ `authors`.
	* \* `require` → `dd/evolutioncms-libraries-ddtools`: Renamed from `dd/modxevo-library-ddtools`.


## Version 2.4.2 (2020-05-06)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.32 is required.
* \* Compatibility with (MODX)EvolutionCMS.libraries.ddTools >= 0.32.


## Version 2.4.1 (2020-05-04)
* \+ README.
* \+ README_ru.
* \+ CHANGELOG.
* \+ CHANGELOG_ru.
* \+ Composer.json.
* \* Refactoring.


## Version 2.4 (2017-12-09)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.21 is required.
* \+ Added an ability to exlude HTML tags from snippet processing (see the `excludeTags` parameter).
* \* Content of the `<code>` tag is excluded by default from the snippet processing (the `excludeTags` parameter by default is equal to `notg,code`).
* \* Minor changes: Code style and comments.


## Version 2.3 (2015-08-24)
* \* PHP.libraries.EMT has been updated to 3.5:
	* \+ Unicode support has been added. At last!
	* \+ The `+X (XXX)XXXXXXX` phone format is now supported. Phones matching this pattern will be rewritten as `+X XXX XXX-XX-XX`.
	* \+ Thin space (`&thinsp;`) is now being used as the thousands separator (`10&thinsp;000`).
	* \* Mistaken transformation of an uppercase domain zone to initials has been fixed (`XXX.RU John Doe` → `XXX.R. U. John Doe`).
	* \* The error occurring while processing nested tags has been fixed.
	* \* Processing of links in quotes has been improved.
	* \* Putting dots in initials has been improved.
	* \* The quotes parsing rule was greatly improved.
	* \* The `$`, `€`, and `№` characters will be complemented with a non-breaking space (`&nbsp;`) either after or before depending on the position of the adjacent number.
* \* The `Nobr.phone_builder` typograph option now depends on the `noTags` snippet parameter.
* \* Forbidding line breaking using HTML tags is now controlled by the only `Etc.nobr_to_nbsp` typograph option.
* \* Other minor changes.


## Version 2.2 (2014-05-25)
* \* Attention! (MODX)EvolutionCMS.libraries.ddTools >= 0.12 is required.
* \* PHP.libraries.EMT has been updated to 3.3:
	* \* The period placement rule has been improved.
	* \* Optical margin alignment is now on by default. Also, the opening quotation mark alignment was improved.
	* \* The Volt symbol bug has been fixed.
	* \* Inch and closing quotation marks will no longer be mixed up.
* \+ The `noTags` parameter has been added. It allows disabling of HTML-tags inserting to the text.
* \- The redundant parameter `Nobr_hyphenNowrap` has been removed. You can use `noTags` instead. Please, contact us if you think that the removed parameter was useful and it should be returned.
* \* The following parameters have been renamed (the snippet works with the old names but they are deprecated):
	* `OptAlign` → `optAlign`.
	* `Text_paragraphs` → `text_paragraphs`.
	* `Text_autoLinks` → `text_autoLinks`.
	* `Etc_unicodeConvert` → `etc_unicodeConvert`.


## Version 2.1 (2014-04-16)
* \+ The `Nobr_hyphenNowrap` parameter, allowing hyphenated words to be rendered in nowrap mode, has been added.


## Version 2.0.2b (2014-03-26)
* \* The `else` statement was added in the string length check to prevent returning void. The snippet now returns the same passed string if its length is less than four symbols.


## Version 2.0.1b (2014-02-25)
* \* The path to PHP.libraries.EMT has been fixed.


## Version 2.0b (2014-02-21)
* \* PHP.libraries.EMT 3.2 is used.
* \- The following parameters have been removed:
	* \- `disableTof`.
	* \- `disableBaseParam`.
* \+ The following parameters have been added:
	* \+ `OptAlign`.
	* \+ `Text_paragraphs`.
	* \+ `Text_autoLinks`.
	* \+ `Etc_unicodeConvert`.
* \* Warning! This version can be capable with the 1.x versions if the only `text` parameter was passed while 1.x using.


## Version 1.4.3 (2013-05-28)
* \* Tiny bugfix.


## Version 1.4.2 (2013-04-05)
* \* Bugfix for PHP 5.3 (files `/assets/snippets/ddtypograph/Jare/Typograph.php` & `/assets/snippets/ddtypograph/Jare/Typograph/Tof.php`).


## Version 1.4.1 (2013-03-06)
* \+ Parsing of `&#34;` is now correct.


## Version 1.4 (2010-10-10)
* \+ Added disabling base-params for tofs.


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />
<style>ul{list-style:none;}</style>

ddTypograph.php
<?php
/**
 * ddTypograph
 * @version 2.4 (2017-12-09)
 * 
 * @desc Snippet for text typography. The snippet doesn’t use third-party services, also it sends no requests. In other words, everything is performed on your server.
 * 
 * @uses PHP >= 5.4.
 * @uses MODXEvo.libraries.ddTools >= 0.21 {@link http://code.divandesign.biz/modx/ddtools }.
 * @uses EMT lib 3.5 (contains in archive).
 * 
 * @param $text {string} — Text to correct. @required
 * @param $optAlign {0|1} — Optical alignment (hanging punctuation). Default: 0.
 * @param $text_paragraphs {0|1} — Section signs and line breaks insertion. Default: 0.
 * @param $text_autoLinks {0|1} — Marking links (including email ones). Default: 0.
 * @param $etc_unicodeConvert {0|1} — Convert html entities into Unicode (“—” instead of “&mdash;” etc.). Default: 1.
 * @param $noTags {0|1} — Whether HTML element insertion is allowed or not. There are cases when using tags causes the text to be invalid, for example, using the snippet inside of an HTML attribute. Default: 0.
 * @param $excludeTags {string_commaSeparated} — HTML tags which content will be ignored by snippet. Default: 'notg,code'.
 * 
 * @link http://code.divandesign.biz/modx/ddtypograph/2.4
 * 
 * @copyright 2010–2017 DivanDesign {@link http://www.DivanDesign.biz }
 */

//Если есть что типографировать
if (strlen($text) > 4){
	global $ddTypograph;
	
	//Заменим кавычки, вставленные через спец. символы на обычные (а то не обрабатываются в библиотеке)
	$text = str_replace(
		'&#34;',
		'"',
		$text
	);
	
	if (!isset($ddTypograph)){
		//Подключаем EMT типограф
		require_once $modx->getConfig('base_path').'assets/snippets/ddTypograph/EMT.php';
		
		$ddTypograph = new EMTypograph();
	}
	
	//Include MODXEvo.libraries.ddTools
	require_once $modx->getConfig('base_path').'assets/libs/ddTools/modx.ddtools.class.php';
	
	//Для обратной совместимости
	extract(ddTools::verifyRenamedParams(
		$params,
		[
			'optAlign' => 'OptAlign',
			'text_paragraphs' => 'Text_paragraphs',
			'text_autoLinks' => 'Text_autoLinks',
			'etc_unicodeConvert' => 'Etc_unicodeConvert'
		]
	));
	
	//Safe tags
	$excludeTags = isset($excludeTags) ? strtolower($excludeTags) : 'notg,code';
	$excludeTags = explode(',', $excludeTags);
	
	foreach ($excludeTags as $excludeTags_item){
		$excludeTags_item = trim($excludeTags_item);
		
		//We don't need anything with default EMT tag
		if ($excludeTags_item != 'notg'){
			//Wrap <notg>
			$text = str_ireplace(
				[
					//Tag start
					'<'.$excludeTags_item,
					//Tag end
					'</'.$excludeTags_item.'>'
				],
				[
					//Tag start
					'<notg><'.$excludeTags_item,
					//Tag end
					'</'.$excludeTags_item.'></notg>'
				],
				$text
			);
		}
	}
	
	//Если нельзя добавлять теги к тексту
	if (
		isset($noTags) &&
		$noTags == 1
	){
// 		$noTags = 'off';
		
		$optAlign = 'off';
		$text_paragraphs = 'off';
		$text_autoLinks = 'off';
		
		$etc_nobr_to_nbsp = 'on';
	}else{
// 		$noTags = 'on';
		
		$optAlign = isset($optAlign) && $optAlign == 1 ? 'on' : 'off';
		$text_paragraphs = isset($text_paragraphs) && $text_paragraphs == 1 ? 'on' : 'off';
		$text_autoLinks = isset($text_autoLinks) && $text_autoLinks == 1 ? 'on' : 'off';
		
		$etc_nobr_to_nbsp = 'off';
	}
	
	$etc_unicodeConvert = isset($etc_unicodeConvert) && $etc_unicodeConvert == 0 ? 'off' : 'on';
	
	$ddTypograph->setup([
		//Расстановка «кавычек-елочек» первого уровня
		'Quote.quotes' => 'on',
		//Внутренние кавычки-лапки
		'Quote.quotation' => 'on',
		
		//Автоматическая простановка дефисов в обезличенных местоимениях и междометиях
		'Dash.to_libo_nibud' => 'on',
		//Расстановка дефисов между «из-за», «из-под»
		'Dash.iz_za_pod' => 'on',
		//Расстановка дефисов перед «-ка», «-де», «-кась».
		'Dash.ka_de_kas' => 'on',
		
		//Привязка союзов и предлогов к написанным после словам
		'Nobr.super_nbsp' => 'on',
		//Привязка союзов и предлогов к предыдущим словам в случае конца предложения
		'Nobr.nbsp_in_the_end' => 'on',
		//TODO: работает плоховато (в «+7 777 777 77 77» ставит неразнывные пробелы только в двух первых случаях), обсудить с Евгением
		//Объединение в неразрывные конструкции номеров телефонов
// 		'Nobr.phone_builder' => $noTags,
		'Nobr.phone_builder' => 'on',
		//Дополнительный формат номеров телефонов («+7(123)1234567» → «+7 123 123-45-67»)
// 		'Nobr.phone_builder_v2' => $noTags,
		'Nobr.phone_builder_v2' => 'on',
		//Объединение IP-адресов.
		'Nobr.ip_address' => 'off',
		//Привязка инициалов к фамилиям («Иванов И. И.» → «Иванов&nbsp;И.&nbsp;И.»)
// 		'Nobr.spaces_nobr_in_surname_abbr' => $noTags,
		'Nobr.spaces_nobr_in_surname_abbr' => 'on',
		//Расстановка точек у инициалов («Иванов И И» | «Иванов ИИ» → «Иванов И. И.»)
// 		'Nobr.dots_for_surname_abbr' => $noTags,
		'Nobr.dots_for_surname_abbr' => 'on',
		//TODO: Не работает (по крайней мере, не удалось увидеть работу)
		//Привязка градусов к числу
		'Nobr.nbsp_celcius' => 'on',
		//TODO: Параметр не ясен
		//Обрамление пятисимвольных слов разделенных дефисом в неразрывные блоки
		'Nobr.hyphen_nowrap_in_small_words' => 'off',
		//Отмена переноса слова с дефисом
// 		'Nobr.hyphen_nowrap' => $noTags,
		'Nobr.hyphen_nowrap' => 'on',
		//TODO: Тег «nobr» невалидный, а для «word-spacing» нет значения «nowrap», нужно использовать свойство «white-space».
		//Использовать nowrap для неразрывных конструкций
		'Nobr.nowrap' => 'on',
		
		//Замена «(tm)» на символ торговой марки «™»
		'Symbol.tm_replace' => 'on',
		//Замена «(r)» на символ зарегистрированной торговой марки «®»
		'Symbol.r_sign_replace' => 'on',
		//Замена «(c)» на символ копирайта «©»
		'Symbol.copy_replace' => 'on',
		//Расстановка правильного апострофа в текстах
		'Symbol.apostrophe' => 'on',
		//TODO: Параметр не ясен
		//Градусы по Фаренгейту
		'Symbol.degree_f' => 'on',
		//TODO: Не срабатывает в конце предложения
		//Замена стрелок «<-» и «->» на символы «←» и «→»
		'Symbol.arrows_symbols' => 'on',
		//TODO: Параметр не ясен
		//Расстановка дюйма после числа
		'Symbol.no_inches' => 'on',
		
		//Расстановка запятых перед «а» и «но»
		'Punctmark.auto_comma' => 'on',
		//Замена трех точек на знак многоточия («...» → «…»)
		'Punctmark.hellip' => 'on',
		//Замена сдвоенных знаков препинания на одинарные
		'Punctmark.fix_pmarks' => 'on',
		//Замена восклицательного и вопросительного знаков местами
		'Punctmark.fix_excl_quest_marks' => 'on',
		//Точка в конце текста, если её там нет
		'Punctmark.dot_on_end' => 'off',
		
		//Расстановка знака минус между числами
		'Number.minus_between_nums' => 'on',
		//TODO: Параметр не ясен
		//Расстановка знака минус между диапозоном чисел
		'Number.minus_in_numbers_range' => 'off',
		//Замена «x» (и по-русски и по-английски) на символ «×» в размерных единицах
		'Number.auto_times_x' => 'on',
		//Замена дробей на соответствующие символы («1/2» → «½», «1/4» → «⅓», «3/4» → «¼»)
		'Number.simple_fraction' => 'off',
		//Математические знаки больше или равно/меньше или равно/плюс минус/неравно («>=» → «≥», «<=» → «≤», «+-» → «±», «!=» → «≠»)
		'Number.math_chars' => 'on',
		//Объединение триад чисел полупробелом (не разбивает на триады, просто заменяет обычный пробел на полупробел)
		'Number.thinsp_between_number_triads' => 'on',
		//Пробел между символом номера и числом
		'Number.thinsp_between_no_and_number' => 'on',
		//Пробел между символом параграфа и числом
		'Number.thinsp_between_sect_and_number' => 'on',
		
		//TODO: Параметр не ясен
		//Установка тире и пробельных символов в периодах дат
		'Date.years' => 'on',
		//Расстановка тире и объединение в неразрывные периоды месяцев
		'Date.mdash_month_interval' => 'off',
		//Расстановка тире и объединение в неразрывные периоды дней
		'Date.nbsp_and_dash_month_interval' => 'off',
		//Привязка года к дате (« 01.01.2015г.» → « 01.01.2015&nbsp;г.»)
		'Date.nobr_year_in_date' => 'on',
		
		//Удаление лишних пробельных символов и табуляций
		'Space.many_spaces_to_one' => 'on',
		//Удаление пробела перед символом процента
		'Space.clear_percent' => 'on',
		//Удаление пробелов перед и после знаков препинания в предложении
		'Space.clear_before_after_punct' => 'on',
		//Расстановка пробелов после знаков препинания
		'Space.autospace_after' => 'on',
		//Удаление пробелов внутри скобок, а также расстановка пробела перед скобками
		'Space.bracket_fix' => 'on',
		
		//Форматирование денежных сокращений (расстановка пробелов и привязка названия валюты к числу)
		'Abbr.nbsp_money_abbr' => 'on',
		//Объединение сокращений «и т. д.», «и т. п.», «в т. ч.»
// 		'Abbr.nobr_vtch_itd_itp' => $noTags,
		'Abbr.nobr_vtch_itd_itp' => 'on',
		//Расстановка пробелов перед сокращениями «см.», «им.»
		'Abbr.nobr_sm_im' => 'on',
		//Расстановка пробелов перед сокращениями «гл.», «стр.», «рис.», «илл.», «ст.», «п.»
		'Abbr.nobr_acronym' => 'on',
		//Расстановка пробелов в сокращениях «г.», «ул.», «пер.», «д.»
		'Abbr.nobr_locations' => 'on',
		//Расстановка пробелов перед сокращениями «dpi», «lpi»
		'Abbr.nobr_abbreviation' => 'on',
		//Объединение сокращений «P.S.», «P.P.S.»
// 		'Abbr.ps_pps' => $noTags,
		'Abbr.ps_pps' => 'on',
		//Привязка сокращений форм собственности к названиям организаций
		'Abbr.nbsp_org_abbr' => 'on',
		//Привязка аббревиатуры «ГОСТ» к номеру
// 		'Abbr.nobr_gost' => $noTags,
		'Abbr.nobr_gost' => 'on',
		//Установка пробельных символов в сокращении вольт
		'Abbr.nobr_before_unit_volt' => 'on',
		//Замена символов и привязка сокращений в размерных величинах («м», «см», «м2», …)
		'Abbr.nbsp_before_unit' => 'on',
		
		//TODO: Разобраться, что это за параметр и какие значения он может принимать
		//Все настройки оптического выравнивания
// 		'OptAlign.all' => 'off',
		//Оптическое выравнивание открывающей кавычки
		'OptAlign.oa_oquote' => $optAlign,
		//Оптическое выравнивание для пунктуации (скобка и запятая)
		'OptAlign.oa_obracket_coma' => $optAlign,
		//TODO: Параметр не ясен
		//Оптическое выравнивание кавычки
		'OptAlign.oa_oquote_extra' => $optAlign,
		//Inline стили или CSS-классы
		'OptAlign.layout' => 'style',
		
		//Простановка параграфов
		'Text.paragraphs' => $text_paragraphs,
		//Выделение ссылок из текста
		'Text.auto_links' => $text_autoLinks,
		//Выделение электронной почты из текста
		'Text.email' => $text_autoLinks,
		//Простановка переносов строк
		'Text.breakline' => $text_paragraphs,
		//Удаление повторяющихся слов
		'Text.no_repeat_words' => 'off',
		
		//Преобразовывать html-сущности в юникод
		'Etc.unicode_convert' => $etc_unicodeConvert,
		//Использовать символ «&nbsp;» вместо тегов для связывания
		'Etc.nobr_to_nbsp' => $etc_nobr_to_nbsp,
		//Разбиение числа на триады («10000» → «10&thinsp;000»)
		'Etc.split_number_to_triads' => 'on'
	]);
	
	$ddTypograph->set_text($text);
	
	//Типографируем
	return $ddTypograph->apply();
}else{
	return $text;
}
?>

src/Snippet.php
<?php
namespace ddTypograph;

class Snippet extends \DDTools\Snippet {
	protected
		$version = '2.5.0',
		
		$params = [
			//Defaults
			'text' => '',
			'optAlign' => false,
			'text_paragraphs' => false,
			'text_autoLinks' => false,
			'etc_unicodeConvert' => true,
			'noTags' => false,
			'excludeTags' => 'notg,code'
		],
		
		$paramsTypes = [
			'optAlign' => 'boolean',
			'text_paragraphs' => 'boolean',
			'text_autoLinks' => 'boolean',
			'etc_unicodeConvert' => 'boolean',
			'noTags' => 'boolean'
		],
		
		$renamedParamsCompliance = [
			'optAlign' => 'OptAlign',
			'text_paragraphs' => 'Text_paragraphs',
			'text_autoLinks' => 'Text_autoLinks',
			'etc_unicodeConvert' => 'Etc_unicodeConvert'
		]
	;
	
	/**
	 * prepareParams
	 * @version 1.0 (2021-03-29)
	 * 
	 * @param $params {stdClass|arrayAssociative|stringJsonObject|stringQueryFormatted}
	 * 
	 * @return {void}
	 */
	protected function prepareParams($params = []){
		//Call base method
		parent::prepareParams($params);
		
		$this->params->excludeTags = explode(
			',',
			strtolower($this->params->excludeTags)
		);
	}
	
	/**
	 * run
	 * @version 1.0 (2021-03-29)
	 * 
	 * @return {string}
	 */
	public function run(){
		//The snippet must return an empty string even if result is absent
		$result = $this->params->text;
		
		//Если есть что типографировать
		if (strlen($result) > 4){
			global $ddTypograph;
			
			//Заменим кавычки, вставленные через спец. символы на обычные (а то не обрабатываются в библиотеке)
			$result = str_replace(
				'&#34;',
				'"',
				$result
			);
			
			if (!isset($ddTypograph)){
				$ddTypograph = new \EMTypograph();
			}
			
			//Safe tags
			foreach (
				$this->params->excludeTags as
				$excludeTags_item
			){
				$excludeTags_item = trim($excludeTags_item);
				
				//We don't need anything with default EMT tag
				if ($excludeTags_item != 'notg'){
					//Wrap <notg>
					$result = str_ireplace(
						[
							//Tag start
							'<' . $excludeTags_item,
							//Tag end
							'</' . $excludeTags_item . '>'
						],
						[
							//Tag start
							'<notg><' . $excludeTags_item,
							//Tag end
							'</' . $excludeTags_item . '></notg>'
						],
						$result
					);
				}
			}
			
			//Если нельзя добавлять теги к тексту
			if ($this->params->noTags){
// 				$noTags = 'off';
				
				$this->params->optAlign = 'off';
				$this->params->text_paragraphs = 'off';
				$this->params->text_autoLinks = 'off';
				
				$etc_nobr_to_nbsp = 'on';
			}else{
//				$noTags = 'on';
				
				$this->params->optAlign =
					$this->params->optAlign ?
					'on' :
					'off'
				;
				$this->params->text_paragraphs =
					$this->params->text_paragraphs ?
					'on' :
					'off'
				;
				$this->params->text_autoLinks =
					$this->params->text_autoLinks ?
					'on' :
					'off'
				;
				
				$etc_nobr_to_nbsp = 'off';
			}
			
			$this->params->etc_unicodeConvert =
				$this->params->etc_unicodeConvert ?
				'on' :
				'off'
			;
			
			$ddTypograph->setup([
				//Расстановка «кавычек-елочек» первого уровня
				'Quote.quotes' => 'on',
				//Внутренние кавычки-лапки
				'Quote.quotation' => 'on',
				
				//Автоматическая простановка дефисов в обезличенных местоимениях и междометиях
				'Dash.to_libo_nibud' => 'on',
				//Расстановка дефисов между «из-за», «из-под»
				'Dash.iz_za_pod' => 'on',
				//Расстановка дефисов перед «-ка», «-де», «-кась».
				'Dash.ka_de_kas' => 'on',
				
				//Привязка союзов и предлогов к написанным после словам
				'Nobr.super_nbsp' => 'on',
				//Привязка союзов и предлогов к предыдущим словам в случае конца предложения
				'Nobr.nbsp_in_the_end' => 'on',
				//TODO: работает плоховато (в «+7 777 777 77 77» ставит неразнывные пробелы только в двух первых случаях), обсудить с Евгением
				//Объединение в неразрывные конструкции номеров телефонов
// 				'Nobr.phone_builder' => $noTags,
				'Nobr.phone_builder' => 'on',
				//Дополнительный формат номеров телефонов («+7(123)1234567» → «+7 123 123-45-67»)
// 				'Nobr.phone_builder_v2' => $noTags,
				'Nobr.phone_builder_v2' => 'on',
				//Объединение IP-адресов.
				'Nobr.ip_address' => 'off',
				//Привязка инициалов к фамилиям («Иванов И. И.» → «Иванов&nbsp;И.&nbsp;И.»)
// 				'Nobr.spaces_nobr_in_surname_abbr' => $noTags,
				'Nobr.spaces_nobr_in_surname_abbr' => 'on',
				//Расстановка точек у инициалов («Иванов И И» | «Иванов ИИ» → «Иванов И. И.»)
// 				'Nobr.dots_for_surname_abbr' => $noTags,
				'Nobr.dots_for_surname_abbr' => 'on',
				//TODO: Не работает (по крайней мере, не удалось увидеть работу)
				//Привязка градусов к числу
				'Nobr.nbsp_celcius' => 'on',
				//TODO: Параметр не ясен
				//Обрамление пятисимвольных слов разделенных дефисом в неразрывные блоки
				'Nobr.hyphen_nowrap_in_small_words' => 'off',
				//Отмена переноса слова с дефисом
// 				'Nobr.hyphen_nowrap' => $noTags,
				'Nobr.hyphen_nowrap' => 'on',
				//TODO: Тег «nobr» невалидный, а для «word-spacing» нет значения «nowrap», нужно использовать свойство «white-space».
				//Использовать nowrap для неразрывных конструкций
				'Nobr.nowrap' => 'on',
				
				//Замена «(tm)» на символ торговой марки «™»
				'Symbol.tm_replace' => 'on',
				//Замена «(r)» на символ зарегистрированной торговой марки «®»
				'Symbol.r_sign_replace' => 'on',
				//Замена «(c)» на символ копирайта «©»
				'Symbol.copy_replace' => 'on',
				//Расстановка правильного апострофа в текстах
				'Symbol.apostrophe' => 'on',
				//TODO: Параметр не ясен
				//Градусы по Фаренгейту
				'Symbol.degree_f' => 'on',
				//TODO: Не срабатывает в конце предложения
				//Замена стрелок «<-» и «->» на символы «←» и «→»
				'Symbol.arrows_symbols' => 'on',
				//TODO: Параметр не ясен
				//Расстановка дюйма после числа
				'Symbol.no_inches' => 'on',
				
				//Расстановка запятых перед «а» и «но»
				'Punctmark.auto_comma' => 'on',
				//Замена трех точек на знак многоточия («...» → «…»)
				'Punctmark.hellip' => 'on',
				//Замена сдвоенных знаков препинания на одинарные
				'Punctmark.fix_pmarks' => 'on',
				//Замена восклицательного и вопросительного знаков местами
				'Punctmark.fix_excl_quest_marks' => 'on',
				//Точка в конце текста, если её там нет
				'Punctmark.dot_on_end' => 'off',
				
				//Расстановка знака минус между числами
				'Number.minus_between_nums' => 'on',
				//TODO: Параметр не ясен
				//Расстановка знака минус между диапозоном чисел
				'Number.minus_in_numbers_range' => 'off',
				//Замена «x» (и по-русски и по-английски) на символ «×» в размерных единицах
				'Number.auto_times_x' => 'on',
				//Замена дробей на соответствующие символы («1/2» → «½», «1/4» → «⅓», «3/4» → «¼»)
				'Number.simple_fraction' => 'off',
				//Математические знаки больше или равно/меньше или равно/плюс минус/неравно («>=» → «≥», «<=» → «≤», «+-» → «±», «!=» → «≠»)
				'Number.math_chars' => 'on',
				//Объединение триад чисел полупробелом (не разбивает на триады, просто заменяет обычный пробел на полупробел)
				'Number.thinsp_between_number_triads' => 'on',
				//Пробел между символом номера и числом
				'Number.thinsp_between_no_and_number' => 'on',
				//Пробел между символом параграфа и числом
				'Number.thinsp_between_sect_and_number' => 'on',
				
				//TODO: Параметр не ясен
				//Установка тире и пробельных символов в периодах дат
				'Date.years' => 'on',
				//Расстановка тире и объединение в неразрывные периоды месяцев
				'Date.mdash_month_interval' => 'off',
				//Расстановка тире и объединение в неразрывные периоды дней
				'Date.nbsp_and_dash_month_interval' => 'off',
				//Привязка года к дате (« 01.01.2015г.» → « 01.01.2015&nbsp;г.»)
				'Date.nobr_year_in_date' => 'on',
				
				//Удаление лишних пробельных символов и табуляций
				'Space.many_spaces_to_one' => 'on',
				//Удаление пробела перед символом процента
				'Space.clear_percent' => 'on',
				//Удаление пробелов перед и после знаков препинания в предложении
				'Space.clear_before_after_punct' => 'on',
				//Расстановка пробелов после знаков препинания
				'Space.autospace_after' => 'on',
				//Удаление пробелов внутри скобок, а также расстановка пробела перед скобками
				'Space.bracket_fix' => 'on',
				
				//Форматирование денежных сокращений (расстановка пробелов и привязка названия валюты к числу)
				'Abbr.nbsp_money_abbr' => 'on',
				//Объединение сокращений «и т. д.», «и т. п.», «в т. ч.»
// 				'Abbr.nobr_vtch_itd_itp' => $noTags,
				'Abbr.nobr_vtch_itd_itp' => 'on',
				//Расстановка пробелов перед сокращениями «см.», «им.»
				'Abbr.nobr_sm_im' => 'on',
				//Расстановка пробелов перед сокращениями «гл.», «стр.», «рис.», «илл.», «ст.», «п.»
				'Abbr.nobr_acronym' => 'on',
				//Расстановка пробелов в сокращениях «г.», «ул.», «пер.», «д.»
				'Abbr.nobr_locations' => 'on',
				//Расстановка пробелов перед сокращениями «dpi», «lpi»
				'Abbr.nobr_abbreviation' => 'on',
				//Объединение сокращений «P.S.», «P.P.S.»
// 				'Abbr.ps_pps' => $noTags,
				'Abbr.ps_pps' => 'on',
				//Привязка сокращений форм собственности к названиям организаций
				'Abbr.nbsp_org_abbr' => 'on',
				//Привязка аббревиатуры «ГОСТ» к номеру
// 				'Abbr.nobr_gost' => $noTags,
				'Abbr.nobr_gost' => 'on',
				//Установка пробельных символов в сокращении вольт
				'Abbr.nobr_before_unit_volt' => 'on',
				//Замена символов и привязка сокращений в размерных величинах («м», «см», «м2», …)
				'Abbr.nbsp_before_unit' => 'on',
				
				//TODO: Разобраться, что это за параметр и какие значения он может принимать
				//Все настройки оптического выравнивания
// 				'OptAlign.all' => 'off',
				//Оптическое выравнивание открывающей кавычки
				'OptAlign.oa_oquote' => $this->params->optAlign,
				//Оптическое выравнивание для пунктуации (скобка и запятая)
				'OptAlign.oa_obracket_coma' => $this->params->optAlign,
				//TODO: Параметр не ясен
				//Оптическое выравнивание кавычки
				'OptAlign.oa_oquote_extra' => $this->params->optAlign,
				//Inline стили или CSS-классы
				'OptAlign.layout' => 'style',
				
				//Простановка параграфов
				'Text.paragraphs' => $this->params->text_paragraphs,
				//Выделение ссылок из текста
				'Text.auto_links' => $this->params->text_autoLinks,
				//Выделение электронной почты из текста
				'Text.email' => $this->params->text_autoLinks,
				//Простановка переносов строк
				'Text.breakline' => $this->params->text_paragraphs,
				//Удаление повторяющихся слов
				'Text.no_repeat_words' => 'off',
				
				//Преобразовывать html-сущности в юникод
				'Etc.unicode_convert' => $this->params->etc_unicodeConvert,
				//Использовать символ «&nbsp;» вместо тегов для связывания
				'Etc.nobr_to_nbsp' => $etc_nobr_to_nbsp,
				//Разбиение числа на триады («10000» → «10&thinsp;000»)
				'Etc.split_number_to_triads' => 'on'
			]);
			
			$ddTypograph->set_text($result);
			
			//Типографируем
			$result = $ddTypograph->apply();
		}
		
		return $result;
	}
}